File: ios/Sources/VaultSync/Domain/UseCases/GetDocumentsUseCase.swift
import Foundation

public final class GetDocumentsUseCase: Sendable {
    private let repository: any DocumentRepository

    public init(repository: any DocumentRepository) {
        self.repository = repository
    }

    public func execute(query: String = "", filterStatus: SyncStatus? = nil) -> AsyncStream<[Document]> {
        let baseStream = repository.observeDocuments()
        return AsyncStream { continuation in
            Task {
                for await documents in baseStream {
                    let filtered = documents.filter { doc in
                        let matchesQuery = query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                            doc.name.localizedCaseInsensitiveContains(query)
                        let matchesStatus = filterStatus == nil || doc.syncStatus == filterStatus
                        return matchesQuery && matchesStatus
                    }
                    continuation.yield(filtered)
                }
                continuation.finish()
            }
        }
    }

    public func executeDirect() async throws -> [Document] {
        try await repository.getDocuments()
    }
}

File: android/app/src/main/java/com/vaultsync/domain/usecase/GetDocumentsUseCase.kt
package com.vaultsync.domain.usecase

import com.vaultsync.domain.model.Document
import com.vaultsync.domain.model.SyncStatus
import com.vaultsync.domain.repository.DocumentRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

class GetDocumentsUseCase(
    private val repository: DocumentRepository
) {
    operator fun invoke(query: String = "", filterStatus: SyncStatus? = null): Flow<List<Document>> {
        return repository.observeDocuments().map { list ->
            list.filter { doc ->
                val matchesQuery = query.isBlank() || doc.name.contains(query, ignoreCase = true)
                val matchesStatus = filterStatus == null || doc.syncStatus == filterStatus
                matchesQuery && matchesStatus
            }
        }
    }

    suspend fun executeDirect(): List<Document> {
        return repository.getDocuments()
    }
}
